spring-cloud-function-adapters/spring-cloud-function-adapter-gcp/src/main/java/org/springframework/cloud/function/adapter/gcp/FunctionInvoker.java

```diff
@@ -91,18 +94,38 @@ private <I> Function<Message<I>, Message<byte[]>> lookupFunction() {
 	 */
 	@Override
 	public void service(HttpRequest httpRequest, HttpResponse httpResponse) throws Exception {
+
 		Function<Message<BufferedReader>, Message<byte[]>> function = lookupFunction();
 
 		Message<BufferedReader> message = getInputType() == Void.class || getInputType() == null ? null
 				: MessageBuilder.withPayload(httpRequest.getReader()).copyHeaders(httpRequest.getHeaders()).build();
+
 		Message<byte[]> result = function.apply(message);
 
 		if (result != null) {
+			MessageHeaders headers = result.getHeaders();
 			httpResponse.getWriter().write(new String(result.getPayload(), StandardCharsets.UTF_8));
-			for (Entry<String, Object> header : result.getHeaders().entrySet()) {
-				httpResponse.appendHeader(header.getKey(), header.getValue().toString());
+			for (Entry<String, Object> header : headers.entrySet()) {
+				Object values = header.getValue();
+				if (values instanceof Collection<?>) {
+					StringBuilder sb = new StringBuilder();
+					for (Iterator iterator = ((Collection<?>) values).iterator(); iterator.hasNext();) {
+						sb.append(iterator.next());
+						if (iterator.hasNext()) {
+							sb.append(",");
+						}
+					}
```

Comment:
`String.join()` will simplify this code.

Reply:
@ddixit14 Please replace custom implementation with built-in utility.

Reply:
Here is a variant for using built-in JDK streaming utilities when you are not starting with a `Collection<String>`:

```
if (header.getValue() instanceof Collection<?> values) {
					String joined = values.stream().map(item -> item.toString()).collect(Collectors.joining(","));
					httpResponse.appendHeader(header.getKey(), joined);
				}
```

Note that pattern-matching `instanceof` is a convenience syntax new to Java 17.